lib/fs/fakefs.go

```diff
@@ -162,8 +168,19 @@ func (fs *fakefs) entryForName(name string) *fakeEntry {
 		if entry.entryType != fakeEntryTypeDir {
 			return nil
 		}
+
 		var ok bool
-		entry, ok = entry.children[comp]
+
+		if fs.insens {
+			for _, child := range entry.children {
+				if strings.EqualFold(child.name, comp) {
+					entry = child
+					ok = true
+				}
+			}
```

Comment:
Why not just store stuff lowercased in the first place?

Reply:
Because a) we'd rather `Stat` returned the filename the way it is capitalized, the way it does on the systems in question, and b) lowercase conversion and case folding are surprisingly different in Unicode.

Reply:
Sure, but only the map keys need to be lower, the entries themselves could have capitalized name, and unicode.ToLower is a thing.

Reply:
Even `strings.ToLower` specifically says "all Unicode letters mapped to their lower case" which ought to be correct enough. The thing with Stat is that it specifically returns the name you passed to it, not whatever is stored at the filesystem level. Otherwise our lives would be easier.

Reply:
OK, let me get priorities straight so I have a better idea of what we're trying to achieve.

Please correct me if I'm wrong here, but I somehow got these ideas for basis:
1. we have an issue where we're reliably losing user data
1. `fakefs` is used for testing and doesn't really do anything in production code
3. cycling through the `map` the way I did it is slow, and slow is bad, we want things fast
4. for tests we want to reproduce actual systems' behaviour as accurately as possible, so that we don't later get into the situation where our code passes the test with flying colors but still fails in production

According to the project goals, not losing data is out first and topmost priority. Thus, when choosing between speed/developing convenience (i.e. test speed) and keeping user's data safe we are bound to choose the latter.

I've updated the PR to have real-world-issue example in test. And [here](https://play.golang.org/p/lqKzKotS5G2) is a code snippet that shows there is effectively no way to make this test pass if we use `strings.ToLower`. That's kind of what I meant with my first comment.

I guess what I'm saying is: it's better to make sure `4` above is achieved, even if it means sacrifising `3` (especially in light of `2`). Otherwise, further down the line trying to achieve `1` is a Sisyphean attempt, because we're leaving a gaping loophole at the very first step.

Reply:
Sure. Fakefs is a limited simulation of reality for specific purposes; it's performance is not completely irrelevant but almost. That said, if the aim is to use this to work towards a solution where Syncthing works well on case insensitive filesystems, it should simulate at least the relevant aspects of those systems that we care about. One of the very specific things that case insensitive file systems (or at least the Go APIs on top of them) do is lie about the case of things when you stat them, always returning the name you asked for rather than what is actually stored on disk.

So I don't particularly care how fast or slow your map lookup is. It doesn't matter. But returning an entry with the name `Foo` when the stat call was for `fOO` is something that doesn't happen in reality and is something that I think would be worth doing the same in our simulation. It's the typical example of something that could make our tests pass with flying colors yet break completely in the real world. That's all. :)

Reply:
Point taken. Will work on `stat` part, too (`strings.ToLowerCase` wouldn't help there much, either). Consider this PR a WIP for now. :)

Reply:
Done. Stat behaves the same way `os.Stat` does on OS X (not tested on Windows, don't have it at home).

Reply:
My point still stands, I think you could have the map keys lowercased so you don't need to iterate over the keys

Reply:
Problem is, I couldn't. `strings.ToLower("ΣΊΣΥΦΟΣ")` is NOT equal to  `strings.ToLower("Σίσυφος")`, and these strings do point to the same file on a case-insensitive FS.

Reply:
```go
func unicodeFoldLower(s string) string {
	rs := []rune(s)
	for i := range rs {
		for r := unicode.SimpleFold(rs[i]); r > rs[i]; r = unicode.SimpleFold(r) {
			rs[i] = r
		}
	}
	return string(rs)
}
```
?

Reply:
Wow. Just... wow.

My attempt on this func is 10% slower, has twice the line count, and tends to prefer uppercase.

Is it a good idea to make this exported and put into /lib/util (seeing as how it will obviously be reused for case-insensitive indexing)?

Reply:
Reworked that and added tests for all I could think of.

Reply:
What about the existing `UnicodeLowercase` at https://github.com/syncthing/syncthing/blob/master/lib/fs/folding.go#L13 ? That uses unicode's `ToLower`/`ToUpper`.

Reply:
Right, I've missed it, thanks for pointing out.